In [2]:
import pandas as pd
import sqlite3
import sqlalchemy
from sqlalchemy import create_engine, text

In [3]:
conn = sqlite3.connect("ladder.db")

In [4]:
pd.read_sql("select name from sqlite_master;", con=conn)

name
0               pets
1          employees
2     employees_null
3       transactions
4                yum
5       yum_by_month
6     trans_by_month
7  trans_by_employee

##### 1) Query all the data in the `pets` table.

In [52]:
pd.read_sql("select * from pets;", conn)

name sex  species  age
0        Chloe   F      dog    9
1    Paddinton   M      dog   10
2        Petey   M      dog    7
3        Bella   F      dog    1
4   Glenn Coco   M      cat    6
5       Alanna   F      cat    3
6         Mimi   F      dog    4
7        Midge   F      cat    7
8          Eli   M      dog    8
9        Shuri   F      cat    2
10      Oliver   M      cat    5
11      Dottie   F      cat    3
12      Pinchy   M  lobster    3

##### #2) Query only the first 5 rows of the `pets` table.

In [53]:
pd.read_sql("select * from pets limit 5", conn)

name sex species  age
0       Chloe   F     dog    9
1   Paddinton   M     dog   10
2       Petey   M     dog    7
3       Bella   F     dog    1
4  Glenn Coco   M     cat    6

##### 3) Query only the names and ages of the pets in the `pets` table.

In [54]:
pd.read_sql('select name, age from pets limit 5', conn)

name  age
0       Chloe    9
1   Paddinton   10
2       Petey    7
3       Bella    1
4  Glenn Coco    6

##### 4) Query the pets in the `pets` table, sorted youngest to oldest.

In [83]:
pd.read_sql('select * from pets order by age', conn)

name sex  species  age
0        Bella   F      dog    1
1        Shuri   F      cat    2
2       Alanna   F      cat    3
3       Dottie   F      cat    3
4       Pinchy   M  lobster    3
5         Mimi   F      dog    4
6       Oliver   M      cat    5
7   Glenn Coco   M      cat    6
8        Petey   M      dog    7
9        Midge   F      cat    7
10         Eli   M      dog    8
11       Chloe   F      dog    9
12   Paddinton   M      dog   10

###### 5) Query the pets in the `pets` table alphabetically.

In [82]:
pd.read_sql('select * from pets order by name', conn)

name sex  species  age
0       Alanna   F      cat    3
1        Bella   F      dog    1
2        Chloe   F      dog    9
3       Dottie   F      cat    3
4          Eli   M      dog    8
5   Glenn Coco   M      cat    6
6        Midge   F      cat    7
7         Mimi   F      dog    4
8       Oliver   M      cat    5
9    Paddinton   M      dog   10
10       Petey   M      dog    7
11      Pinchy   M  lobster    3
12       Shuri   F      cat    2

###### 6) Query all the male pets in the `pets` table.

In [81]:
pd.read_sql('select * from pets where sex = "M"', conn)

name sex  species  age
0   Paddinton   M      dog   10
1       Petey   M      dog    7
2  Glenn Coco   M      cat    6
3         Eli   M      dog    8
4      Oliver   M      cat    5
5      Pinchy   M  lobster    3

###### 7) Query all the cats in the `pets` table.

In [80]:
pd.read_sql('select * from pets where species = "cat"', conn)

name sex species  age
0  Glenn Coco   M     cat    6
1      Alanna   F     cat    3
2       Midge   F     cat    7
3       Shuri   F     cat    2
4      Oliver   M     cat    5
5      Dottie   F     cat    3

###### 8) Query all the pets in the `pets` table that are at least 5 years old.

In [79]:
pd.read_sql('select * from pets where age >= 5', conn)

name sex species  age
0       Chloe   F     dog    9
1   Paddinton   M     dog   10
2       Petey   M     dog    7
3  Glenn Coco   M     cat    6
4       Midge   F     cat    7
5         Eli   M     dog    8
6      Oliver   M     cat    5

###### 9) Query all the male dogs in the `pets` table. Do not include the sex or species column, since you already know them.

In [77]:
pd.read_sql('select* from pets where species = "dog" and sex = "M" ', conn)

name sex species  age
0  Paddinton   M     dog   10
1      Petey   M     dog    7
2        Eli   M     dog    8

###### 10) Get all the names of the dogs in the `pets` table that are younger than 5 years old.

In [78]:
pd.read_sql('select* from pets where species = "dog" and age < 5', conn)

name sex species  age
0  Bella   F     dog    1
1   Mimi   F     dog    4

###### 11) Query all the pets in the `pets` table that are either male dogs or female cats.

In [88]:
query = """
select*
from pets
where species = 'dog' and sex = 'M' or species = 'cat' and sex = 'F'
"""
pd.read_sql(query, conn)

name sex species  age
0  Paddinton   M     dog   10
1      Petey   M     dog    7
2     Alanna   F     cat    3
3      Midge   F     cat    7
4        Eli   M     dog    8
5      Shuri   F     cat    2
6     Dottie   F     cat    3

###### 12) Query the five oldest pets in the `pets` table.

In [90]:
query = """
select *
from pets
order by age desc
limit 5
"""
pd.read_sql(query, conn)

name sex species  age
0  Paddinton   M     dog   10
1      Chloe   F     dog    9
2        Eli   M     dog    8
3      Petey   M     dog    7
4      Midge   F     cat    7

###### 13) Get the names and ages of all the female cats in the `pets` table sorted by age, descending.

In [98]:
query = """
select name, age, sex
from pets
where species = 'cat' and sex = 'F'
order by age desc
"""
pd.read_sql(query, conn)

name  age sex
0   Midge    7   F
1  Alanna    3   F
2  Dottie    3   F
3   Shuri    2   F

###### 14) Get all pets from `pets` whose names start with P.

In [110]:
query = """
select *
from pets
where name like 'P%'
"""
pd.read_sql(query, conn)

name sex  species  age
0  Paddinton   M      dog   10
1      Petey   M      dog    7
2     Pinchy   M  lobster    3

###### 15) Select all employees from `employees_null` where the salary is missing.

In [116]:
query = """
select *
from employees_null
where salary is NULL
"""
pd.read_sql(query, conn)

ID firstname lastname            job salary   startdate
0  22     Maria     Owen          Sales   None  2010-03-22
1  47   Shannon   Rivera  Administrator   None  2007-09-12
2  53  Jennifer     Cruz          Sales   None  2015-11-09
3  56   Russell     Rice          Sales   None  2008-09-02
4  59     Andre   Levine          Sales   None  2016-03-27
5  60      Anna  Fischer          Sales   None  1994-12-25
6  69  Danielle      Orr          Sales   None  2004-06-24
7  79    Tamara  Douglas          Sales   None  2017-10-03
8  85   Candice   Wright          Sales   None  1998-03-04
9  98   Bradley   Romero  Administrator   None  2009-05-10

###### 16) Select all employees from `employees_null` where the salary is below $35,000 or missing.

In [118]:
query = """
select *
from employees_null
where salary < 35000
"""

pd.read_sql(query, conn)

ID firstname  lastname    job  salary   startdate
0   26  Samantha   Nichols  Sales   33248  2015-04-20
1   55    Julian  Martinez  Sales   34597  2009-06-19
2   65   Michael      West  Sales   32781  1991-08-08
3   73    Miguel  Williams  Sales   31584  2004-04-02
4  101   UNKNOWN   UNKNOWN   None      -1        None

###### 17) Select all employees from `employees_null` where the job title is missing. What do you see?

In [125]:
query = """
select *
from employees_null
where job IS NULL
"""

pd.read_sql(query, conn)

ID firstname lastname   job  salary startdate
0  101   UNKNOWN  UNKNOWN  None      -1      None

###### 18) Who is the newest employee in `employees`? The most senior?

In [129]:
# most senior employee
query = """
Select *
from employees
order by startdate asc
limit 1
"""

pd.read_sql(query, conn)

ID firstname lastname    job  salary   startdate
0  32      Mary     Nash  Sales   65472  1990-02-16

In [130]:
# newest employee 
query = """
Select *
from employees
order by startdate desc
limit 1
"""
pd.read_sql(query, conn)

ID firstname lastname    job  salary   startdate
0  19     Roger   Conner  Sales   60375  2019-05-24

###### 19) Select all employees from `employees` named Thomas.

In [135]:
query = """
Select *
from employees
where firstname = 'Thomas'
"""
pd.read_sql(query, conn)

ID firstname  lastname    job  salary   startdate
0   1    Thomas      Peck  Sales  112972  2011-01-04
1  45    Thomas    Miller  Sales   44836  2002-11-17
2  80    Thomas  Gonzalez  Sales   84772  2019-01-05

###### 20) Select all employees from `employees` named Thomas or Shannon.

In [143]:
query = """
Select *
from employees
where firstname = 'Thomas' or firstname = 'Shannon'
"""
pd.read_sql(query, conn)

ID firstname  lastname            job  salary   startdate
0   1    Thomas      Peck          Sales  112972  2011-01-04
1  33   Shannon    Bailey          Sales   66071  2014-12-30
2  45    Thomas    Miller          Sales   44836  2002-11-17
3  47   Shannon    Rivera  Administrator   82791  2007-09-12
4  80    Thomas  Gonzalez          Sales   84772  2019-01-05

###### 21) Select all employees from `employees` named Robert, Lisa, or any name that begins with a J. In addition, only show employees who are _not_ in sales. This will be a little bit of a longer query.

In [204]:
query = """
Select *
from employees
where firstname = 'Thomas' or firstname = 'Lisa' or firstname like 'J%'
EXCEPT
select*
from employees
Where  job = 'Sales'
"""
pd.read_sql(query, conn)

ID firstname lastname            job  salary   startdate
0   4    Janice   Martin             IT   62007  2011-02-25
1   6      Lisa   Cooper     Operations   56265  1994-02-20
2   8      Jose  Rosario  Administrator  111905  2008-05-05
3  39     Jenny   Barber             IT  110679  2019-01-10
4  72    Joseph   Campos     Operations   88345  1992-06-10

###### 22) Query the top 5 rows of the `employees` table to get a glimpse of these new data.

In [205]:
query = """
select*
from employees
limit 5
"""
pd.read_sql(query, conn)

ID    firstname  lastname    job  salary   startdate
0   0    Christine  Thompson  Sales  123696  2005-01-20
1   1       Thomas      Peck  Sales  112972  2011-01-04
2   2  Christopher    Robles     IT   78426  2003-12-10
3   3    Elizabeth     Munoz  Sales   55824  1993-07-28
4   4       Janice    Martin     IT   62007  2011-02-25

###### 23) Query the `employees` table, but convert their salaries to Euros. 
    * _Hint:_ 1 Euro = 1.1 USD.
    * _Hint2:_ If you think the output is ugly, try out the `ROUND()` function.

In [217]:
query = """
select 
    ID,
    firstname,
    lastname,
    job,
    startdate,
    salary * 1.1 as salary_eur
from employees
limit 5
"""
pd.read_sql(query, conn)

ID    firstname  lastname    job   startdate  salary_eur
0   0    Christine  Thompson  Sales  2005-01-20    136065.6
1   1       Thomas      Peck  Sales  2011-01-04    124269.2
2   2  Christopher    Robles     IT  2003-12-10     86268.6
3   3    Elizabeth     Munoz  Sales  1993-07-28     61406.4
4   4       Janice    Martin     IT  2011-02-25     68207.7

###### 24) Repeat the previous problem, but rename the column `salary_eu`.

In [222]:
query = """
select 
    ID,
    firstname,
    lastname,
    job,
    salary * 1.1 as salary_eu, 
    startdate
from employees
limit 5
"""
pd.read_sql(query, conn)

ID    firstname  lastname    job  salary_eu   startdate
0   0    Christine  Thompson  Sales   136065.6  2005-01-20
1   1       Thomas      Peck  Sales   124269.2  2011-01-04
2   2  Christopher    Robles     IT    86268.6  2003-12-10
3   3    Elizabeth     Munoz  Sales    61406.4  1993-07-28
4   4       Janice    Martin     IT    68207.7  2011-02-25

###### 25) Query the `employees` table, but combine the `firstname` and `lastname` columns to be "Firstname, Lastname" format. Call this column `fullname`. For example, the first row should contain `Thompson, Christine` as `fullname`. Also, display the rounded `salary_eu` instead of `salary`.
    * _Hint:_ The string concatenation operator is `||`

In [231]:
query = """
select 
    ID,
    firstname ||', '|| lastname as fullname,
    job,
    round(salary * 1.1, 2) as salary_eu, 
    startdate
from employees
limit 5
"""
pd.read_sql(query, conn)

ID             fullname    job  salary_eu   startdate
0   0  Christine, Thompson  Sales   136065.6  2005-01-20
1   1         Thomas, Peck  Sales   124269.2  2011-01-04
2   2  Christopher, Robles     IT    86268.6  2003-12-10
3   3     Elizabeth, Munoz  Sales    61406.4  1993-07-28
4   4       Janice, Martin     IT    68207.7  2011-02-25

###### 26) Query the `employees` table, but replace `startdate` with `startyear` using the `SUBSTR()` function. Also include `fullname` and `salary_eu`.

In [234]:
query = """
select 
    firstname ||', '|| lastname as fullname,
    round(salary * 1.1, 2) as salary_eu, 
    SUBSTR(startdate, 1, INSTR(startdate, '-') - 1) AS startyear
from employees
limit 5
"""
pd.read_sql(query, conn)

fullname  salary_eu startyear
0  Christine, Thompson   136065.6      2005
1         Thomas, Peck   124269.2      2011
2  Christopher, Robles    86268.6      2003
3     Elizabeth, Munoz    61406.4      1993
4       Janice, Martin    68207.7      2011

###### 27) Repeat the above problem, but instead of using `SUBSTR()`, use `STRFTIME()`.

In [235]:
query = """
select 
    firstname ||', '|| lastname as fullname,
    round(salary * 1.1, 2) as salary_eu, 
    STRFTIME('%Y', startdate) AS startyear
from employees
limit 5
"""
pd.read_sql(query, conn)

fullname  salary_eu startyear
0  Christine, Thompson   136065.6      2005
1         Thomas, Peck   124269.2      2011
2  Christopher, Robles    86268.6      2003
3     Elizabeth, Munoz    61406.4      1993
4       Janice, Martin    68207.7      2011

###### 28) Query the `employees` table, replacing `firstname`/`lastname` with `fullname` and `startdate` with `startyear`. Print out the salary in USD again, except format it with a dollar sign, comma separators, and no decimal. For example, the first row should read `$123,696`. This column should still be named `salary`.
    * _Hint:_ Check out SQLite's `printf` function.
    * _Hint2:_ The format string you'll need is `$%,.2d`. You should read more about such formatting strings as they're useful in Python, too!

In [238]:
query = """
select 
    firstname ||', '|| lastname as fullname,
    printf('$%,.2d', salary) as salary,
    STRFTIME('%Y', startdate) AS startyear
from employees
limit 5
"""
pd.read_sql(query, conn)

fullname    salary startyear
0  Christine, Thompson  $123,696      2005
1         Thomas, Peck  $112,972      2011
2  Christopher, Robles   $78,426      2003
3     Elizabeth, Munoz   $55,824      1993
4       Janice, Martin   $62,007      2011

**Note:** For the next few problems, you'll probably want to use `CASE`/`WHEN` statements.

###### 29) Last year, only salespeople were eligible for bonuses. Create a column `bonus` that is "Yes" if you're eligible for a bonus, otherwise "No".

In [17]:
query = """
select

    firstname ||', '|| lastname as fullname,
    job,
    printf('$%,.2d', salary) as salary,
    STRFTIME('%Y', startdate) AS startyear,
    case when job = 'Sales' then 'Yes' else 'No' END as bonus
from employees
limit 5
"""
pd.read_sql(query, conn)

fullname    job    salary startyear bonus
0  Christine, Thompson  Sales  $123,696      2005   Yes
1         Thomas, Peck  Sales  $112,972      2011   Yes
2  Christopher, Robles     IT   $78,426      2003    No
3     Elizabeth, Munoz  Sales   $55,824      1993   Yes
4       Janice, Martin     IT   $62,007      2011    No

###### 30) This year, only sales people with a salary of $100,000 or higher are eligible for bonuses. Create a bonus column like in the last problem for salespeople with salaries at least $100,000.

In [18]:
query = """
select 
    firstname ||', '|| lastname as fullname,
    job,
    printf('$%,.2d', salary) as salary,
    STRFTIME('%Y', startdate) AS startyear,
    case when salary >= 100000 then 'Yes' else 'No' END as bonus
from employees
where salary >= 100000 
order by salary asc
limit 5
"""
pd.read_sql(query, conn)

fullname    job    salary startyear bonus
0    Andrew, Larson  Sales  $100,095      2003   Yes
1  Briana, Calderon  Sales  $101,218      1993   Yes
2      Cindy, Allen  Sales  $101,923      2018   Yes
3   Joseph, Russell  Sales  $102,572      2006   Yes
4  Natasha, Michael  Sales  $106,288      2008   Yes

###### 31) Next year, the bonus structure will be a little more complicated. You'll create a `target_comp` column which represents an employee's target total compensation after their bonus. Here is the company's bonus structure:

* Salespeople who make more than $100,000 will be eligible for a 10% bonus.
* Salespeople who make less than $100,000 will be eligible for a 5% bonus.
* Administrators will also be eligible for a 5% bonus.
* Anyone who does not meet any of the above descriptions is not eligible for a bonus.

Create this `target_comp` column, making sure to format _both_ the `salary` and `target_comp` columns nicely (ie, with dollar signs and comma separators).

In [72]:
query = """
select 
fullname,
job,
salary,
startyear,
printf('$%,.2d', target_comp) as target_comp

from (select 
    firstname ||', '|| lastname as fullname,
    job,
    printf('$%,.2d', salary) as salary,
    STRFTIME('%Y', startdate) AS startyear,
    case 
    when salary >= 100000 then salary + (salary * 0.1) 
    when salary <= 100000 then salary + (salary * 0.05) END as target_comp   
from employees
where job = 'Sales'
limit 10)
"""
pd.read_sql(query, conn)

fullname    job    salary startyear target_comp
0  Christine, Thompson  Sales  $123,696      2005    $136,065
1         Thomas, Peck  Sales  $112,972      2011    $124,269
2     Elizabeth, Munoz  Sales   $55,824      1993     $58,615
3         Kevin, Young  Sales   $44,886      2011     $47,130
4     Steve, Robertson  Sales   $43,384      2007     $45,553
5         Cindy, Allen  Sales  $101,923      2018    $112,115
6     Veronica, Hughes  Sales   $63,788      2018     $66,977
7         Cory, Gaines  Sales   $93,346      2008     $98,013
8        Roger, Conner  Sales   $60,375      2019     $63,393
9         Lisa, Morgan  Sales  $120,555      1996    $132,610